### PR TITLE
Added more fields to Consultation Form

### LINKED ISSUE #859: More Fields to be added in Consultation Api
Part of [#2218](https://github.com/coronasafe/care_fe/issues/2218)
2 collapsable sections would be added to the consultation form **Health Section** and **Medical History** extra fields need to be added as the information given in the issue [#2218](https://github.com/coronasafe/care_fe/issues/2218)

### PR DESCRIPTION
**Pull request by cp-Coder**
# Updates

- [x] Fixes #859
- [x] Related to coronasafe/care_fe#2218, coronasafe/care_fe#2738
- [x] Added fields required in coronasafe/care_fe#3332
- [x] Added fields required in coronasafe/care_fe#3333
- [x] Added fields required in coronasafe/care_fe#3335
- [x] Added fields required in coronasafe/care_fe#3336
- [x] Added fields required in coronasafe/care_fe#3338
- [x] Created a new Patient Health Details, model
- [x] Added Allergies, Blood Group & Family History in Patient Health Details
- [x] Added serializer, viewset, and API endpoint for Patient Health Details

### CODE REVIEW

**Comment by gigincg** [comment]
Waiting for FE before merging

**Comment by cp-Coder** [comment]
image

File: care/facility/api/serializers/patient_health_details.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    id = serializers.CharField(source="external_id", read_only=True)
+
+    patient = ExternalIdSerializerField(
+        queryset=PatientRegistration.objects.all()
+    )
+    facility = ExternalIdSerializerField(queryset=Facility.objects.all())
+    facility_object = FacilityBasicInfoSerializer(
+        source="facility", read_only=True
+    )
+    created_in_consultation = ExternalIdSerializerField(
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_health_details.py:42]: we can just call it consultation

File: care/facility/api/serializers/patient_health_details.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                "last_vaccinated_date",
+            )
+
+    id = serializers.CharField(source="external_id", read_only=True)
+
+    patient = ExternalIdSerializerField(
+        queryset=PatientRegistration.objects.all()
+    )
+    facility = ExternalIdSerializerField(queryset=Facility.objects.all())
+    facility_object = FacilityBasicInfoSerializer(
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_health_details.py:39]: why is this required?

File: care/facility/api/serializers/patient_health_details.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from care.facility.models.patient import Vaccine
+from care.facility.models.patient_base import (
+    BLOOD_GROUP_CHOICES,
+    VACCINE_CHOICES,
+)
+from care.utils.serializer.external_id_field import ExternalIdSerializerField
+
+
+class PatientHealthDetailsSerializer(serializers.ModelSerializer):
+    class VaccinationHistorySerializer(serializers.Serializer):
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_health_details.py:20]: shouldn't this also be a model serializer?

Comment (vigneshhari) [inline care/facility/api/serializers/patient_health_details.py:20]: define the serializer outside, so that we can reuse it at some point

File: care/facility/api/serializers/patient_consultation.py [hunk 4]
```diff
@@ -239,9 +302,50 @@ def create(self, validated_data):
...
+        try:
+            health_details_data = self.context["request"].data[
+                "new_health_details"
+            ]
+            vaccination_history = health_details_data.pop(
+                "vaccination_history", []
+            )
+
+            health_details = PatientHealthDetails(
+                patient=consultation.patient,
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:314]: All of these should be authorised

File: care/facility/api/serializers/patient_consultation.py [hunk 5]
```diff
@@ -165,8 +184,46 @@ def update(self, instance, validated_data):
...
+        try:
+            health_details_data = self.context["request"].data[
+                "new_health_details"
+            ]
+            vaccination_history = health_details_data.pop(
+                "vaccination_history", []
+            )
+            PatientHealthDetails.objects.filter(
+                id=consultation.last_health_details.id
+            ).update(**self.context["request"].data["new_health_details"])
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:196]: use validated data, do not use raw data from the request

File: care/facility/api/serializers/patient_consultation.py [hunk 6]
```diff
@@ -165,8 +184,46 @@ def update(self, instance, validated_data):
...
+            consultation.last_health_details = (
+                PatientHealthDetails.objects.get(
+                    id=consultation.last_health_details.id
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:199]: No Auth Present

File: care/facility/api/serializers/patient_consultation.py [hunk 7]
```diff
@@ -165,8 +184,46 @@ def update(self, instance, validated_data):
...
+            vaccines = []
+            for vaccine in vaccination_history:
+                vaccines.append(
+                    Vaccine(
+                        health_details=consultation.last_health_details,
+                        **vaccine,
+                    )
+                )
+            if vaccines:
+                Vaccine.objects.bulk_create(vaccines, ignore_conflicts=True)
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:211]: does bulk create handle changes in one of the fields?

File: care/facility/api/serializers/patient_consultation.py [hunk 8]
```diff
@@ -165,8 +184,46 @@ def update(self, instance, validated_data):
...
+            consultation.save(update_fields=["last_health_details"])
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:212]: the id does not change, right? so do we need to save the consultation again?

File: care/facility/api/serializers/patient_consultation.py [hunk 9]
```diff
@@ -165,8 +184,46 @@ def update(self, instance, validated_data):
...
+            NotificationGenerator(
+                event=Notification.Event.PATIENT_HEALTH_DETAILS_UPDATED,
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient_consultation.py:214]: I think we can ignore this notification, @gigincg the consultation notification should be enough right?

File: care/facility/migrations/0310_auto_20220813_1123.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            has_allergy = True
+            allergies = patient.allergies
+
+        if patient.blood_group is not None:
+            blood_group = patient.blood_group
+
+        health_details_objs.append(
+            HealthDetails(
+                patient=patient,
+                facility=patient.facility,
```
Comment (vigneshhari) [inline care/facility/migrations/0310_auto_20220813_1123.py:25]: facility should be consultation's facility

**Review by vigneshhari — CHANGES_REQUESTED** [review]

**Comment by vigneshhari** [comment]
Let's not delete any fields immediately; let's keep a leeway of 1 month before deleting any data. remove any migration that deletes the columns.

File: care/facility/api/serializers/patient_consultation.py (hunk 8 above)
Comment (cp-Coder) [reply care/facility/api/serializers/patient_consultation.py:212]: @vigneshhari In the request, there is a `new_health_details` field through which we are creating a `health_details` record. After this, we have to display this record as the `last_health_details` field in the consultation API. So we are basically initializing the field with the record after the consultation object has been created. That's why I am saving the consultation object again (only doing a partial update)

File: care/facility/api/serializers/patient_consultation.py (hunk 5 above)
Comment (cp-Coder) [reply care/facility/api/serializers/patient_consultation.py:196]: @vigneshhari I am using the value from `new_health_details` (not a field in consultation API) to create a `health_detail` record. The `validated_data` will remove the `new_health_details` field that's why I am passing the data this way.

File: care/facility/api/serializers/patient_consultation.py (hunk 4 above)
Comment (cp-Coder) [reply care/facility/api/serializers/patient_consultation.py:314]: @vigneshhari The authorization checks are already present in the consultation API for consultation. Will the auth check for `health_details` be different from the consultation?

File: care/facility/api/serializers/patient_consultation.py (hunk 9 above)
Comment (cp-Coder) [reply care/facility/api/serializers/patient_consultation.py:226]: I have talked with Gigin. He told me to remove this notification

**Comment by cp-Coder** [comment]
Will open a new PR squashing the commits
